crates/turborepo-lib/src/task_graph/visitor.rs

```diff
@@ -78,10 +109,40 @@ impl<'a> Visitor<'a> {
                 _ => (),
             }
 
-            let _task_def = engine
+            let task_definition = engine
                 .task_definition(&info)
                 .ok_or(Error::MissingDefinition)?;
 
+            let task_env_mode = match self.global_env_mode {
+                // Task env mode is only independent when global env mode is `infer`.
+                EnvMode::Infer if !task_definition.pass_through_env.is_empty() => {
+                    ResolvedEnvMode::Strict
+                }
+                // If we're in infer mode we have just detected non-usage of strict env vars.
+                // But our behavior's actual meaning of this state is `loose`.
+                EnvMode::Infer => ResolvedEnvMode::Loose,
+                // Otherwise we just use the global env mode.
+                EnvMode::Strict => ResolvedEnvMode::Strict,
+                EnvMode::Loose => ResolvedEnvMode::Loose,
+            };
+
+            let dependency_set = engine.dependencies(&info).ok_or(Error::MissingDefinition)?;
+
+            let task_hash = self
+                .task_hasher
+                .calculate_task_hash(
+                    &self.scm,
+                    &info,
+                    &task_definition,
+                    task_env_mode,
+                    workspace_info,
+                    dependency_set,
+                    self.repo_root,
+                )
+                .await?;
```

Comment:
This will end up blocking the processing of any other tasks by the visitor since it happens in the main thread and not via one of the `tokio::spawn` tasks.